src/tools/cephfs/DataScan.cc

```diff
@@ -0,0 +1,1267 @@
+
+  inodeno_t ino = backtrace.ino;
+  dout(10) << "  inode: 0x" << std::hex << ino << std::dec << dendl;
+  for (std::vector<inode_backpointer_t>::const_iterator i = backtrace.ancestors.begin();
+      i != backtrace.ancestors.end(); ++i) {
+    const inode_backpointer_t &backptr = *i;
+    dout(10) << "  backptr: 0x" << std::hex << backptr.dirino << std::dec
+      << "/" << backptr.dname << dendl;
+
+    // Examine root dirfrag for parent
+    const inodeno_t parent_ino = backptr.dirino;
+    const std::string dname = backptr.dname;
+
+    frag_t fragment;
+    int r = get_frag_of(parent_ino, dname, &fragment);
+    if (r == -ENOENT) {
+      // Don't know fragment, fall back to assuming root
+      dout(20) << "don't know fragment for 0x" << std::hex <<
+        parent_ino << std::dec << "/" << dname << ", will insert to root"
+        << dendl;
+
+      // TODO: if we can't find it by following backtrace back by inos,
+      // we *could* try finding it by following dnames forwards
+      // from the root, but that only works if we're lucky and the
+      // backtrace is up to date.  Might not be worth implementing,
+      // if we can just use live injection (i.e. via a running MDCache)
+      // for that case.
+  
+      // TODO: limited-size cache of directory fragtrees, so that when
+      // injecting lots of files to a dir we don't keep on doing
+      // the lookup process to learn its fragtree
+    }
+
+    // Find or create dirfrag
+    // ======================
+    bool created_dirfrag;
+    r = find_or_create_dirfrag(parent_ino, fragment, &created_dirfrag);
+    if (r < 0) {
+      return r;
+    }
+
+    // TODO: if backtrace does not end at root, and the most distance ancestor
+    // did not exist, then link the most distant ancestor into lost+found
+
+    // Check if dentry already exists
+    // ==============================
+    InodeStore existing_dentry;
+    r = read_dentry(parent_ino, fragment, dname, &existing_dentry);
+    bool write_dentry = false;
+    if (r == -ENOENT || r == -EINVAL) {
+      // Missing or corrupt dentry
+      write_dentry = true;
+    } else if (r < 0) {
+      derr << "Unexpected error reading dentry 0x" << std::hex
+        << parent_ino << std::dec << "/"
+        << dname << ": " << cpp_strerror(r) << dendl;
+      break;
+    } else {
+      // Dentry already present, does it link to me?
+      if (existing_dentry.inode.ino == ino) {
+        dout(20) << "Dentry 0x" << std::hex
+          << parent_ino << std::dec << "/"
+          << dname << " already exists and points to me" << dendl;
+        // TODO: a tool or frag to overwrite the 'size' in the
+        // existing metadata with the one that we have just recovered
+        // from scan_extents phase
+      } else {
+        // FIXME: at this point we should set a flag to recover
+        // this inode in a /_recovery/<inodeno>.data file as we
+        // can't recover it into its desired filesystem position.
```

Comment:
For a stray usually recovering into lost+found is still the right thing.

There is the case where a hardlink exists, pointing to a stray, and the inodes backtrace points to the stray, but the stray dentry doesn't exist -- in that case we should restore to the stray dir instead of lost+found.  However, knowing the hardlink exists isn't doable from this tool, we need to get a tip-off from the forward scrub somehow, to tell us when it has seen a broken hardlink.  Or we need that particular case of hardlink repair to be something that only online repairs can do.